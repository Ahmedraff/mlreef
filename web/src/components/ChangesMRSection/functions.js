import { Base64 } from 'js-base64';
import Base64ToArrayBuffer from 'base64-arraybuffer';
import { getFileDifferences } from '../../functions/apiCalls';
import CommitsApi from 'apis/CommitsApi';
import { isImageFormat } from 'functions/dataParserHelpers';

const commitsApi = new CommitsApi();

export const fetchFileDiffInfo = (projectId, aheadCommits) => Promise.all(aheadCommits.map(async (commit, index) => {
  const { id: commitId } = commit;
  const commitDiffs = await commitsApi.getCommitDiff(projectId, commitId, 0, false);
  return await Promise.all(commitDiffs?.body?.map(async (commitDiff) => {
    const {
      previousVersionFile,
      nextVersionFile,
      imageFileSize,
    } = await getFileDifferences(projectId, commitDiff, commit.parent_ids[0], commit.id);
    let previousVersionFileParsed = previousVersionFile;
    let nextVersionFileParsed = nextVersionFile;
    let sizeDeleted = 0;
    let sizeAdded = 0;
    const fileName = commitDiff.old_path.split('/').slice(-1)[0];
    if (previousVersionFile && !isImageFormat(fileName)) {
      previousVersionFileParsed = Base64.decode(
        Base64ToArrayBuffer.encode(previousVersionFile),
      );
      sizeDeleted = previousVersionFile.byteLength;
    }
    if (nextVersionFile && !isImageFormat(fileName)) {
      nextVersionFileParsed = Base64.decode(Base64ToArrayBuffer.encode(nextVersionFile));
      sizeAdded = nextVersionFile.byteLength;
    }
    const path = commitDiff.new_path
      ? commitDiff.new_path
      : commitDiff.old_path;
    const fileExtension = path.split('.').pop().toLowerCase();
    return {
      fileName,
      fileExtension,
      previousVersionFileParsed,
      nextVersionFileParsed,
      id: `${commit.id} ${index}`,
      sizeDeleted,
      sizeAdded,
      imageFileSize,
    };
  }));
}));
