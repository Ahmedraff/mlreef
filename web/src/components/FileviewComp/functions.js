import FilesApi from 'apis/FilesApi.ts';
import CommitsApi from 'apis/CommitsApi.ts';

const commitsApi = new CommitsApi();

const filesApi = new FilesApi();

const getCommit = (projectId, lastCommitId) => commitsApi
  .getCommitDetails(projectId, lastCommitId);

const getFileAndInformation = (gid, file, branch, commit) => filesApi
  .getFileData(gid, file?.includes('/') ? encodeURIComponent(file) : file, branch || commit)
  .then(async (fData) => {
    const commitInfoDet = await getCommit(gid, fData.last_commit_id);
    return {
      fData,
      commitInfoDet,
    };
  });

export default {
  getFileAndInformation,
};
