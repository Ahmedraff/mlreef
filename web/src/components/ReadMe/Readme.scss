/* ---------------------------- Readme section ----------------------------*/
@import '../../styles/breakpoints';

$dark-red: #c0341d;
$light-red: #fbe5e1;

.readme-container {
    border-top: 0;
    border-radius: 4px;
    background-color: var(--white);
    border: 1px solid var(--almostWhite);
    margin: 1rem 0;
  }
  
  .readme-titlebar {
    display: flex;
    flex-direction: row;
    align-items: center;
    margin: 0px auto;
    border-radius: 0.3vw 0.3vw 0 0;
    background-color: var(--almostWhite);
    border-bottom: 1px solid var(--lessWhite);
  }
  
  .readme-profile-pic {
    height: 2em;
    width: 2em;
    background-color: var(--white);
    border: 1px solid var(--lessWhite);
    border-radius: 0.3vw;
    margin-left: 3%;
    display: flex;
  }
  
  .readme-name {
    margin-left: 2.5%;
    color: var(--dark);
  }

  .file-preview {
    max-width: 990px;
    margin-left: auto;
    margin-right: auto;
    background: var(--white);
    font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, "Noto Sans", Ubuntu, Cantarell, "Helvetica Neue", sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol", "Noto Color Emoji";
    font-weight: 400;
    line-height: 1.5;
    text-align: left;

    &.md {
     padding: 2rem; 
    }
  }

  .md {
    position: relative;
    color: var(--dark);
    font-size: 1.2em;

    & *:first-child {
      margin-top: 0;
    };

    a {
      color: var(--info);
    };

    a.absent {
      color: $dark-red;
    };

    a.anchor {
      display: block;
      padding-left: 30px;
      margin-left: -30px;
      cursor: pointer;
      position: absolute;
      top: 0;
      left: 0;
      bottom: 0;
    }

    p {
      margin: 0 0 16px;
      color: var(--dark);
    };

    ul, ol {
      padding: 0;
      margin: 0 0 1rem;
    };

    li {
      margin-left: 2rem;
      padding-left: 0;
      line-height: 1.6em;
    };

    code {
      font-family: "Menlo", "DejaVu Sans Mono", "Liberation Mono", "Consolas", "Ubuntu Mono", "Courier New", "andale mono", "lucida console", monospace;
      white-space: pre-wrap;
      word-wrap: normal;
      word-break: keep-all;
    };

    p>code {
      font-weight: inherit;
    };

    pre {
      margin-bottom: 1rem;
      font-size: 13px;
      line-height: 1.6em;
      overflow-x: auto;
      border-radius: 2px;
    };

    pre code {
      white-space: pre;
      background-color: inherit;
      padding: unset;
      color: var(--dark)
    };

    blockquote {
      border-left: 4px solid var(--lessWhite);
      padding: 0 15px;
      color: #777777;
    };

    blockquote > :first-child {
      margin-top: 0;
    };

    blockquote > :last-child {
      margin-bottom: 0;
    };

    table {
      padding: 0;
    };

    table tr {
      border-top: 1px solid var(--almostWhite);
      background-color: var(--white);
      margin: 0;
      padding: 0;
    };

    table tr th {
      font-weight: bold;
      border: 1px solid var(--almostWhite);
      text-align: left;
      margin: 0;
      padding: 6px 13px;
    };

    table tr td {
      border: 1px solid var(--almostWhite);
      text-align: left;
      margin: 0;
      padding: 6px 13px;
    };

    img {
      max-width: 100%;
    }
  }

  code {
    padding: 2px 4px;
    color: $dark-red;
    background-color: $light-red;
    border-radius: 4px;
    font-size: 1rem;
  }

  pre {
    font-family: "Menlo", "DejaVu Sans Mono", "Liberation Mono", "Consolas", "Ubuntu Mono", "Courier New", "andale mono", "lucida console", monospace;
    display: block;
    padding: 0.5rem 0.75rem;
    margin: 0 0 0.5rem;
    word-break: break-all;
    word-wrap: break-word;
    color: var(--dark);
    background-color: var(--white);
    border: 1px solid var(--almostWhite);
    border-radius: 2px;
  }
