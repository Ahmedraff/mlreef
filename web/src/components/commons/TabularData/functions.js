import genUUID from 'uuid/v1';

const parseNumber = (text) => {
  const num = parseFloat(text);

  return Number.isNaN(num) ? text : num;
};

export const transpose = (matrix) => matrix[0].map((_, ci) => matrix.map((row) => row[ci]));

const sanatizeMatrix = (matrix) => {
  const head = matrix[0];
  const tail = matrix[matrix.length - 1];

  const sameLength = head.length === tail.length;
  const lastRowfilled = tail.every((col) => col !== undefined);

  return sameLength && lastRowfilled ? matrix : matrix.slice(0, -1);
};

export const cvsToArray = (text, separator = ',') => {
  const matrix = text.split('\n')
    .map((line) => line.split(separator)
      .map(parseNumber));

  return sanatizeMatrix(matrix);
};

export const jsonToArray = (input) => {
  const items = Array.isArray(input) ? input : Object.values(input);

  if (!items.length) return items;
  // const row = Object.values(input)[0];
  const row = items[0];
  const header = Object.keys(row).map((label) => label);

  // const data = Object.values(input)
  const data = items
    .map((cols) => Object.values(cols).map((val) => val));

  const matrix = [header].concat(data);

  return sanatizeMatrix(matrix);
};

export const arrayToRichData = (matrix, separator = ',') => {
  const header = matrix[0];

  return {
    id: genUUID(),
    meta: {
      separator,
      timestamp: new Date(),
    },
    matrix,
    // data: flat(matrix.map((row, y) => row.map((value, x) => ({ x, y, value }))))
    data: matrix.map((cols, y) => ({
      id: y,
      cols: cols.map((value, x) => ({
        x,
        y,
        value,
        label: header[x],
      })),
    })),
  };
};
