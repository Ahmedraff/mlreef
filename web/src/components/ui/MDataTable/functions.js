const extractField = (row, x) => row.cols.find((col) => col.x === x)?.value;

export const sortDesc = (x) => (rows) => rows.slice()
  .sort((a, b) => extractField(b, x) - extractField(a, x));

export const sortAsc = (x) => (rows) => rows.slice()
  .sort((a, b) => extractField(a, x) - extractField(b, x));

export const round = (decs) => (input) => {
  const decimals = parseInt(decs, 10);

  if (Number.isNaN(decimals)) return input;

  if (decimals > 12) return input;

  const num = parseFloat(input);
  if (Number.isNaN(num)) return input;

  return Math.round(num * (10 ** decimals)) / (10 ** decimals);
};
