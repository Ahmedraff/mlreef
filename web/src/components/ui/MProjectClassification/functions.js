import { ALGORITHM, DATA_OPERATION, MODEL, OPERATION, VISUALIZATION } from "dataTypes";

export const getSearchable = (feSearchableTypes) => {
  switch (feSearchableTypes) {
    case DATA_OPERATION:
      return OPERATION;
    case MODEL:
      return ALGORITHM;
    default:
      return VISUALIZATION;
  }
}