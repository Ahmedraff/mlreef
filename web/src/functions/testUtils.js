import { createStore, applyMiddleware } from 'redux';

import rootReducer from '../reducers/index';
import { middlewares } from '../configureStore';

// eslint-disable-next-line import/prefer-default-export
export const storeFactory = (initialState) => {
  const createStoreWithMiddleware = applyMiddleware(...middlewares)(createStore);
  return createStoreWithMiddleware(rootReducer, initialState);
};
