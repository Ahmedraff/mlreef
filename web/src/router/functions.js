/**
 * Compose like: f.g x = g(f(x))
 *
 * @example compose (fn4, fn3, fn2, fn1) (arg) === fn4(fn3(fn2(fn1(arg))))
 *
 * @param {Array[Function]} fns sequence of functions.
 * @param arg input parameter.
 *
 * @return far left funcion return.
 */
export const c = (...fns) => (arg) => fns.reverse().reduce((acc, fn) => fn(acc), arg);

/**
 * Return the recorded route that matches name.
 */
export const findRouteInfo = (routes) => ({ name }) => routes.find((route) => route.name === name);

/**
 * Replace placeholders by parameters. e.g. /:user/messages -> /nick/messages.
 *
 * @param {Object(Route)} to implicit.
 * @param {Object} params e.g. key-val object, from the to object.
 * @param {Object(Route)} record the route from routes.
 *
 * @return {Object(Route)} route with path modified.
 */
export const replaceParams = (to) => (record = {}) => {
  if (to.params) {
    const path = Object.entries(to.params)
      .reduce((final, [k, v]) => final.replace(new RegExp(`:${k}`), v), record.path || '');

    return { ...record, path };
  }

  return record;
};

/**
 * Add the hash part. e.g. #last-menu-item
 *
 * @param {Object(Route)} to implicit.
 * @param {String} hash e.g. last-menu-item, from the to object.
 * @param {Object(Route)} record the route from routes.
 *
 * @return {Object(Route)} route with path modified.
 */
export const addHash = (to) => (record) => !to.hash ? record
  : { ...record, path: `${record.path}#${to.hash}` };

export const makeQuery = (query) => Object.entries(query)
  .map(([k, v]) => `${k}=${v}`)
  .join('&');

/**
 * Add the query search. e.g. ?user=nickname&priority=high
 *
 * @param {Object(Route)} to implicit.
 * @param {Object} query e.g. key-val object, from the to object.
 * @param {Object(Route)} record the route from routes.
 *
 * @return {Object(Route)} route with path modified.
 */
export const addQuery = (to) => (record) => !to.query ? record
  : { ...record, path: `${record.path}?${makeQuery(to.query)}` };

export const unpack = ({ path }) => path;

export const resolveRoute = (routes, to) => c(
  unpack, addQuery(to), addHash(to), replaceParams(to), findRouteInfo(routes),
)(to);
