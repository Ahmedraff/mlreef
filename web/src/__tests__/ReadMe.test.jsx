import React from 'react';
import { mount } from 'enzyme';
import Readme from 'components/ui/ReadMe/ReadMe';
import { generatePromiseResponse, sleep } from 'functions/testUtils';

const base64ReadmeContent = 'IyBTaWduIExhbmd1YWdlIENsYXNzaWZpZXIgUmVwbwoKRGVzY3JpcHRpb24KCgpqYWphamFqYQoKCmpqYWphamFhamEgMg==';

const setup = () => mount(
  <Readme
    projectId={1}
    branch="master"
  />,
);

describe('render basic readme elements', () => {
  let wrapper;
  beforeEach(() => {
    jest.spyOn(global, 'fetch')
      .mockImplementation(() => generatePromiseResponse(
        200,
        true,
        { content: base64ReadmeContent },
        50,
      ));
    wrapper = setup();
  });
  test('assert that snapshot matches', async () => {
    await sleep(200);
    wrapper.setProps({});
    const markDownRenderer = wrapper.find('ReactMarkdownWithHtml');
    expect(markDownRenderer.find('h1').childAt(0).text()).toBe('Sign Language Classifier Repo');

    const textRenderers = markDownRenderer.find('TextRenderer');
    expect(textRenderers.at(1).text()).toBe('Description');
    expect(textRenderers.at(2).text()).toBe('jajajaja');
  });
});
