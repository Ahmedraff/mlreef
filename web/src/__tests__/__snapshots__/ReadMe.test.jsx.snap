// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render basic readme elements assert that snapshot matches 1`] = `
<div
  className="readme-container"
>
  <div
    className="readme-titlebar"
  >
    <div
      className="readme-profile-pic"
    />
    <div
      className="readme-name"
    >
      <p>
        <b>
          README.md
        </b>
      </p>
    </div>
  </div>
  <div
    className="file-preview md"
  />
</div>
`;
